File: selenium/src/main/java/pages/IncomePage.java
package pages;

import base.BasePage;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

public class IncomePage extends BasePage {

    private final By addIncomeButton = By.xpath("//main//button[contains(., 'Add Income')]");
    private final By amountInput = By.cssSelector("form input[type='number']");
    private final By categorySelect = By.cssSelector("form select");
    private final By descriptionInput = By.cssSelector("form input[placeholder='Source of income']");
    private final By notesTextarea = By.cssSelector("form textarea");
    private final By submitButton = By.cssSelector("form button[type='submit']");
    private final By searchBox = By.cssSelector("input[placeholder='Search income...']");
    private final By toastSuccess = By.xpath("//*[@role='status' and contains(., 'successfully')]");

    public IncomePage(WebDriver driver) {
        super(driver);
    }

    public void navigateTo(String baseUrl) {
        driver.get(baseUrl + "/income");
    }

    public void addIncome(String amount, String category, String description, String notes) {
        click(addIncomeButton);
        type(amountInput, amount);
        
        selectOptionContainingText(categorySelect, category);

        type(descriptionInput, description);
        if (notes != null && !notes.isEmpty()) {
            type(notesTextarea, notes);
        }
        click(submitButton);
        waitForElementVisible(toastSuccess);
    }
    
    public void search(String query) {
        type(searchBox, query);
    }

    public boolean isSuccessToastDisplayed() {
        return isDisplayed(toastSuccess);
    }

    private void selectOptionContainingText(By selectLocator, String text) {
        WebElement matchingOption = wait.until(driver -> {
            Select currentSelect = new Select(waitForElementVisible(selectLocator));
            for (WebElement option : currentSelect.getOptions()) {
                if (text.equals(option.getAttribute("value")) || option.getText().contains(text)) {
                    return option;
                }
            }
            return null;
        });

        Select select = new Select(waitForElementVisible(selectLocator));
        if (text.equals(matchingOption.getAttribute("value"))) {
            select.selectByValue(text);
            return;
        }
        if (matchingOption.getText().contains(text)) {
            select.selectByVisibleText(matchingOption.getText());
            return;
        }
        throw new IllegalArgumentException("No option containing text: " + text);
    }
}

File: e2e/pages/IncomePage.js
export class IncomePage {
  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;
    this.addIncomeButton = page.locator('main').getByRole('button', { name: 'Add Income' });
    this.amountInput = page.locator('form input[type="number"]');
    this.categorySelect = page.locator('form select').first();
    this.descriptionInput = page.locator('form input[placeholder="Source of income"]');
    this.notesTextarea = page.locator('form textarea');
    this.submitButton = page.locator('form button[type="submit"]');
    
    this.searchBox = page.locator('input[placeholder="Search income..."]');
    this.tableRows = page.locator('tbody tr');
    
    this.toastSuccess = page.getByRole('status').filter({ hasText: /successfully/i });
  }

  async goto() {
    await this.page.goto('/income');
  }

  async addIncome(amount, category, description, notes = '') {
    await this.addIncomeButton.click();
    await this.amountInput.fill(amount.toString());
    await this.categorySelect.selectOption(category);
    await this.descriptionInput.fill(description);
    if (notes) await this.notesTextarea.fill(notes);
    await this.submitButton.click();
    await this.page.locator('form').waitFor({ state: 'detached', timeout: 10000 });
  }

  async searchIncome(term) {
    await this.searchBox.fill(term);
  }
}
